core/src/main/java/com/linecorp/armeria/client/logging/AbstractLoggingClient.java

```diff
@@ -114,14 +116,21 @@
         this.responseContentSanitizer = requireNonNull(responseContentSanitizer, "responseContentSanitizer");
         this.responseTrailersSanitizer = requireNonNull(responseTrailersSanitizer, "responseTrailersSanitizer");
         this.responseCauseSanitizer = requireNonNull(responseCauseSanitizer, "responseCauseSanitizer");
-        this.sampler = requireNonNull(sampler, "sampler");
+        this.successSampler = requireNonNull(successSampler, "successSampler");
+        this.failureSampler = requireNonNull(failureSampler, "failureSampler");
     }
 
     @Override
     public final O execute(ClientRequestContext ctx, I req) throws Exception {
-        if (sampler.isSampled(ctx)) {
```

Comment:
Instead of having 2 new axes, success/failure function then sucess/failure sampler, could sampler just be updated to accept both ctx and log? The default sampler would still be implemented to behave as it does right now

Reply:
You mean like
```
CompletableFututre<Boolean> isSampled(ctx, RequestLog);
```
So the sampler can have ability to check sample or not? Let me check if it can fit in other place also.

Reply:
Checked the source and maybe wanna discuss if possible.  
`Sampler` is public interface with `boolean isSampled(T object);` already, so we may not want to change it to `CompletableFututre<Boolean> isSampled(ctx, RequestLog)`


And @ikhoon Do you think it's good idea to make `SuccessFunction#isSuccess` returns `CompletableFuture<Boolean>` ? Because we may prefer the implementation of `isSuccess` using `requestLog.whenComplete()` directly but not at decorator like `LoggingService`?

Reply:
Just realized that Sampler is exposed and maybe I misunderstood what you mean.

Refactor the code to combine `successSampler`, `failureSampler` and `successFunction` to single Sampler inside LoggingClient/Server